src/backend/javascript-generator.ts

```diff
@@ -0,0 +1,202 @@
+/*
+ * Translation to JavaScript
+ *
+ * Requiring this module adds a gen() method to each of the AST classes.
+ * Nothing is actually exported from this module.
+ *
+ * Each gen() method returns a fragment of JavaScript. The gen() method on
+ * the program class pretty-prints the complete JavaScript code.
+ *
+ *   require('./backend/javascript-generator');
+ *   program.gen();
+ */
+
+import jsBeautify = require('js-beautify');
+
+import Argument from '../ast/argument';
+import AssignmentStatement from '../ast/assignment-statement';
+import BinaryExpression from '../ast/binary-expression';
+import BooleanLiteral from '../ast/boolean-literal';
+import BreakStatement from '../ast/break-statement';
+import Call from '../ast/call';
+import FunctionDeclaration from '../ast/function-declaration';
+import FunctionObject from '../ast/function-object';
+import IdentifierExpression from '../ast/identifier-expression';
+import IfStatement from '../ast/if-statement';
+import ListExpression from '../ast/list-expression';
+import NumericLiteral from '../ast/numeric-literal';
+import Parameter from '../ast/parameter';
+import Program from '../ast/program';
+import ReturnStatement from '../ast/return-statement';
+import StringLiteral from '../ast/string-literal';
+import SubscriptedExpression from '../ast/subscripted-expression';
+import UnaryExpression from '../ast/unary-expression';
+import Variable from '../ast/variable';
+import VariableDeclaration from '../ast/variable-declaration';
+import WhileStatement from '../ast/while-statement';
+import Context from '../semantics/context';
+import { Statement } from '../type-definitions/plainscript-types';
+
+const indentLevel = 2;
+
+const opConversionDictionary: { [key: string]: string } = {
+  '!=': '!==',
+  '==': '===',
+  'and': '&&',
+  'not': '!',
+  'or': '||',
+};
+
+function makeOp(op: string): string {
+  return opConversionDictionary[op] || op;
+}
+
+// jsName(e) takes any PlainScript object with an id property, such as a
+// Variable, Parameter, or FunctionDeclaration, and produces a JavaScript
+// name by appending a unique identifying suffix, such as '_1' or '_503'.
+// It uses a cache so it can return the same exact string each time it is
+// called with a particular entity.
+const jsName = (() => {
+  let lastId = 0;
+  const map = new Map();
+  return (v: Variable | FunctionDeclaration | Parameter) => {
```

Comment:
Cooooooooool

Reply:
yes ! the code became much more self-documented !